src/cryptenroll/cryptenroll-tpm2.c

```diff
@@ -257,8 +257,19 @@ int enroll_tpm2(struct crypt_device *cd,
                         return log_error_errno(SYNTHETIC_ERRNO(ENOTRECOVERABLE), "TPM2 seal/unseal verification failed.");
         }
 
+        if (use_pin) {
+                /* To prevent using a secret from a compromised TPM without the pin, append salted pin.
+                 * See following attack on AMD's fTPMs: https://arxiv.org/abs/2304.14717 */
+                secret_and_salt_size = secret_size + sizeof(salted_pin);
+                secret_and_salt = malloc(secret_and_salt_size);
+                if (!secret_and_salt)
+                        return log_oom();
+                memcpy(secret_and_salt, secret, secret_size);
+                memcpy(secret_and_salt + secret_size, salted_pin, sizeof(salted_pin));
+        }
+
         /* let's base64 encode the key to use, for compat with homed (and it's easier to every type it in by keyboard, if that might end up being necessary. */
-        base64_encoded_size = base64mem(secret, secret_size, &base64_encoded);
+        base64_encoded_size = base64mem(secret_and_salt ?: secret, secret_and_salt_size ?: secret_size, &base64_encoded);
```

Comment:
I am pretty sure this would be more readable, if you'd move this into the `if (use_pin)` check above, and add an else branch. i.e.:

```c
if (use_pin) {
        …
        base64_encoded_size = base64mem(secret_and_salt, secret_and_salt_size, &base64_encoded);
} else 
        base64_encoded_size = base64mem(secret, secret_size, &base64_encoded);
```

i.e. multiple ternary ops that repeatedly check the same conditions (or pretty much the same) suck

Reply:
this also then allows you to reduce the scope for secret_and_salt/secret_and_salt_size